lib/phoenix_live_view/async_result.ex

```diff
@@ -0,0 +1,501 @@
+defmodule Phoenix.LiveView.AsyncResult do
+  @moduledoc ~S'''
+  Adds async functionality to LiveViews and LiveComponents.
+
+  Performing asynchronous work is common in LiveViews and LiveComponents.
+  It allows the user to get a working UI quicky while the system fetches some
+  data in the background or talks to an external service. For async work,
+  you also typically need to handle the different states of the async operation,
+  such as loading, error, and the successful result. You also want to catch any
+  errors or exits and translate it to a meaningful update in the UI rather than
+  crashing the user experience.
+
+  ## Async assigns
+
+  The `assign_async/3` function takes a name, a list of keys which will be assigned
+  asynchronously, and a function that returns the result of the async operation.
+  For example, let's say we want to async fetch a user's organization from the database,
+  as well as their profile and rank:
+
+      def mount(%{"slug" => slug}, _, socket) do
+        {:ok,
+         socket
+         |> assign(:foo, "bar")
+         |> assign_async(:org, fn -> {:ok, %{org: fetch_org!(slug)} end)
+         |> assign_async([:profile, :rank], fn -> {:ok, %{profile: ..., rank: ...}} end)}
+      end
+
+  Here we are assigning `:org` and `[:profile, :rank]` asynchronously. The async function
+  must return a `{:ok, assigns}` or `{:error, reason}` tuple, where `assigns` is a map of
+  the keys passed to `assign_async`. If the function returns other keys or a different
+  set of keys, an error is raised.
+
+  The state of the async operation is stored in the socket assigns within an
+  `%AsyncResult{}`. It carries the loading and error states, as well as the result.
+  For example, if we wanted to show the loading states in the UI for the `:org`,
+  our template could conditionally render the states:
+
+  ```heex
+  <div :if={@org.state == :loading}>Loading organization...</div>
+  <div :if={org = @org.ok? && @org.result}}><%= org.name %> loaded!</div>
+  ```
+
+  The `with_state` function component can also be used to declaratively
+  render the different states using slots:
+
+  ```heex
+  <AsyncResult.with_state :let={org} assign={@org}>
```

Comment:
Assuming we move it apart, I would literally call the component `<.async_result`.